cgmes/cgmes-conversion/src/main/java/com/powsybl/cgmes/conversion/update/Changelog.java

```diff
@@ -0,0 +1,97 @@
+/**
+ * Copyright (c) 2019, RTE (http://www.rte-france.com)
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+package com.powsybl.cgmes.conversion.update;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import com.powsybl.iidm.network.Identifiable;
+import com.powsybl.iidm.network.Network;
+import com.powsybl.iidm.network.NetworkListener;
+
+/**
+ * @author Elena Kaltakova <kaltakovae at aia.es>
+ * @author Luma Zamarreño <zamarrenolm at aia.es>
+ */
+public class Changelog implements NetworkListener {
+
+    /**
+     * Register Network changes
+     *
+     * @param network the IIDM Network
+     */
+    public Changelog(Network network) {
+        network.addListener(this);
+        this.baseChanges = new ArrayList<>();
+        this.changesByVariant = new HashMap<>();
+    }
+
+    @Override
+    public void onCreation(Identifiable identifiable) {
+        baseChanges.add(new IidmChangeCreation(identifiable));
+    }
+
+    @Override
+    public void onRemoval(Identifiable identifiable) {
+        baseChanges.add(new IidmChangeRemoval(identifiable));
+    }
+
+    @Override
+    public void onUpdate(Identifiable identifiable, String attribute, Object oldValue, Object newValue) {
+        if (!IGNORED_ATTRIBUTES.contains(attribute)) {
+            baseChanges.add(new IidmChangeUpdate(identifiable, attribute, oldValue, newValue));
+        }
+    }
+
+    @Override
+    public void onUpdate(Identifiable identifiable, String attribute, String variantId, Object oldValue, Object newValue) {
+        if (!IGNORED_ATTRIBUTES.contains(attribute)) {
+            // Create a new list of changes if no changelog is found for the variant
+            // or if the previous changelog was null
+            changesByVariant.computeIfAbsent(variantId, k -> new ArrayList<>()).add(new IidmChangeUpdate(identifiable, attribute, oldValue, newValue));
+        }
+    }
+
+    @Override
+    public void onVariantCreated(String sourceVariantId, String targetVariantId) {
+        // Will overwrite any previous changelog saved for target variant
+        List<IidmChange> sourceChanges = changesByVariant.get(sourceVariantId);
+        if (sourceChanges != null) {
+            changesByVariant.put(targetVariantId, new ArrayList<>(sourceChanges));
```

Comment:
Why do you wrap (copy) the sourceChanges to a new list?

Reply:
Specific changes by variant are maintained in separate lists. When new variant is created the changelog for the target variant must contain all the changes made in the source up to this point. If additional changes are performed in the source variant after the new variant is created, they must not be present in the target variant changelog, so a copy is required. New changes on the target variant must be added only to the changelog of the target variant.